File: csharp/collatz-conjecture/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        if (number <= 0)
            throw new ArgumentException();

        if (number == 1)
            return 0;

        if (number % 2 == 0)
            return 1 + CollatzConjecture.Steps(number / 2);

        else
            return 1 + CollatzConjecture.Steps(3 * number + 1); 
    }
}

File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture


let addOne (number: int option) : int option =
    match number with
    | Some x -> Some(x + 1)
    | None -> None


let nextNumber (number: int) : int =
    if (number % 2 = 0) then
        (number / 2)
    else
        (number * 3 + 1)


let rec steps (number: int) : int option =
    match number with
    | n when n <= 0 -> None
    | n when n = 1 -> Some 0
    | _ -> number |> nextNumber |> steps |> addOne
